### PR TITLE
[MDS-6086] Updated permit condition extraction to happen asynchronously using Celery

### PR DESCRIPTION
**Pull request by simensma-fresh**
## Objective 

[MDS-6086](https://bcmines.atlassian.net/browse/MDS-6086)

- Added a celery worker for local development that handles permit condition extraction jobs
- Updated permit condition extraction endpoint to create a celery job
- Added endpoints to retrieve the status and results of a permit condition job
- A couple of tweaks to the test reporting scripts to fix a couple of issues that popped up when running it against large permits

image

### CODE REVIEW

**Review by henryoforeh-dev — DISMISSED** [review]

File: services/permits/app/compare_extraction_results.py [hunk 1]
```diff
@@ -193,6 +214,8 @@ def validate_condition(csv_pairs):
             key=lambda c: c.get("Key"),
         )
 
+        print([c.get("Key") for c in match_results["context_comparison_results"]])
```
Comment (taraepp) [inline services/permits/app/compare_extraction_results.py]: this log looks more like it was left in by mistake?

File: services/permits/app/permit_conditions/converters/pdf_to_text_converter.py [hunk 2]
```diff
@@ -62,9 +66,9 @@ def _read_pdf(self, file_path: Path) -> List[str]:
                 for idx, page in enumerate(pdf_reader.pages):
                     try:
                         page_text = page.extract_text(
-                            extraction_mode="layout",
-                            layout_mode_space_vertically=False,
-                            layout_mode_scale_weight=0.5,
+                            # extraction_mode="",
```
Comment (taraepp) [inline services/permits/app/permit_conditions/converters/pdf_to_text_converter.py]: ?

File: services/permits/app/permit_conditions/resources/permit_condition_resource.py [hunk 3]
```diff
@@ -42,56 +50,83 @@ async def extract_permit_conditions(file: UploadFile = File(...)) -> PermitCondi
...
+        The task ID of the job.
+    Raises:
+        400 Bad Request: If the task has not completed successfully
+    """
+    res = run_permit_condition_pipeline.app.AsyncResult(task_id)
 
+    if res.status == "SUCCESS":
+        return res.get()
+    else:
+        raise HTTPException(400, detail=f"Task has not completed successfully. Current status: {res.status}")
```
Comment (taraepp) [inline services/permits/app/permit_conditions/resources/permit_condition_resource.py]: I think I'd expect to get a 400 if I sent a bad task id, but I wonder if there's something more appropriate to send back to indicate "still working on it". Maybe `202 Accepted`? Or a 500 if it's a bad status?

Comment (simensma-fresh) [reply services/permits/app/permit_conditions/resources/permit_condition_resource.py]: Good call, 202 seems good for the "processing" stage of it.
I updated it to be:
`200` -> Task finished
`202` -> Task created, but not finished
`500` -> Task failed

File: services/permits/app/compare_extraction_results.py (hunk 1 above)
Comment (simensma-fresh) [reply services/permits/app/compare_extraction_results.py]: Definitely, removed it.

**Review by henryoforeh-dev — DISMISSED** [review]

File: services/permits/app/permit_conditions/resources/permit_condition_resource.py [hunk 4]
```diff
@@ -42,56 +50,86 @@ async def extract_permit_conditions(file: UploadFile = File(...)) -> PermitCondi
...
+        400 Bad Request: If the task has not completed successfully
+    """
+    res = run_permit_condition_pipeline.app.AsyncResult(task_id)
 
+    if res.status == "SUCCESS":
+        return res.get()
+    elif res.status == "FAILURE":
+        raise HTTPException(500, detail=f"Task failed to complete: {res.status}")
+    else:
+        raise HTTPException(202, detail=f"Task has not completed yet. Current status: {res.status}")
```
Comment (taraepp) [inline services/permits/app/permit_conditions/resources/permit_condition_resource.py]: Does it work to return a 200-type code with HTTPException? If it works, cool, but it looks kinda fishy to me. 😄 

May want to consider an enum with SUCCESS/FAILURE/PROGRESS statuses.

File: services/permits/app/compare_extraction_results.py [hunk 5]
```diff
@@ -103,6 +117,8 @@ def validate_condition(csv_pairs):
         auto_extracted_content = create_content_instances(auto_extracted_df)
         manual_extracted_content = create_content_instances(manual_extracted_df)
 
+        print(f'Found {len(auto_extracted_content)} conditions in {auto_extracted_csv} and {len(manual_extracted_content)} conditions in {manual_extracted_csv}')
```
Comment (asinn134) [inline services/permits/app/compare_extraction_results.py:120]: Was this print statement meant to be added as well?

Comment (simensma-fresh) [reply services/permits/app/compare_extraction_results.py:120]: Indeed it was intentional

File: services/permits/app/permit_conditions/resources/permit_condition_resource.py (hunk 4 above)
Comment (simensma-fresh) [reply services/permits/app/permit_conditions/resources/permit_condition_resource.py]: It does indeed work, but yup not the most clear.

**Comment by simensma-fresh** [comment]
@taraepp I updated the response models (the 202 now shows up in the swagger docs as well) + added an enum. Also added some missing tests.

image

**Review by taraepp — APPROVED** [review]

**Review by asinn134 — APPROVED** [review]

**Pull request merged**